cli/flox-rust-sdk/src/providers/catalog.rs

```diff
@@ -0,0 +1,174 @@
+/// Whether to use an actual catalog client or the mock client.
+///
+/// Don't use a feature flag since this shouldn't be exposed to users.
+static USE_CATALOG_MOCK: Lazy<bool> = Lazy::new(|| env::var("_FLOX_USE_CATALOG_MOCK").is_ok());
+
+/// Either a client for the actual catalog service,
+/// or a mock client for testing.
+#[derive(Debug)]
+pub enum Client {
+    Catalog(CatalogClient),
+    Mock(MockClient),
+}
+
+impl Client {
+    pub fn new(mock: bool) -> Self {
+        if mock {
+            Client::Mock(MockClient)
+        } else {
+            Client::Catalog(CatalogClient::default())
+        }
+    }
+}
+
+impl Default for Client {
+    fn default() -> Self {
+        Self::new(*USE_CATALOG_MOCK)
+    }
+}
+
+/// A client for the catalog service.
+///
+/// This is a wrapper around the auto-generated APIClient.
+#[derive(Debug)]
+pub struct CatalogClient {
+    client: APIClient,
+}
+impl CatalogClient {
+    pub fn new() -> Self {
+        Self {
+            client: APIClient::new(DEFAULT_CATALOG_URL),
+        }
+    }
+}
+impl Default for CatalogClient {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+pub struct MockClient;
+
+impl Client {
+    /// Resolve a list of [PackageGroup]s into a list of
+    /// [ResolvedPackageGroup]s.
+    pub async fn resolve(
+        &self,
+        package_groups: Vec<PackageGroup>,
+    ) -> Result<Vec<ResolvedPackageGroup>, CatalogClientError> {
+        match self {
+            Client::Catalog(client) => {
+                let package_groups = api_types::PackageGroups {
+                    items: package_groups
+                        .into_iter()
+                        .map(TryInto::try_into)
+                        .collect::<Result<Vec<_>, _>>()?,
+                };
+
+                let response = client
+                    .client
+                    .resolve_api_v1_catalog_resolve_post(&package_groups)
+                    .await
+                    .map_err(CatalogClientError::Resolution)?;
+
+                let resolved_package_groups = response.into_inner();
+
+                resolved_package_groups
+                    .items
+                    .into_iter()
+                    .map(TryInto::try_into)
+                    .collect::<Result<Vec<_>, _>>()
+            },
+            Client::Mock(_) => unimplemented!(),
+        }
+    }
+}
+
+/// Just an alias until the auto-generated PackageDescriptor diverges from what
+/// we need.
+pub type PackageDescriptor = api_types::PackageDescriptor;
+
+pub struct PackageGroup {
+    pub descriptors: Vec<PackageDescriptor>,
+    pub name: String,
+    pub system: System,
+}
+
+#[derive(Debug, Error)]
+pub enum CatalogClientError {
+    #[error("system not supported by catalog")]
+    UnsupportedSystem(#[source] api_error::ConversionError),
+    #[error("resolution failed")]
+    Resolution(#[source] APIError<api_types::ErrorResponse>),
+}
```

Comment:
I guess we could be a bit more fine grained here (considering that we have to support most of what is currently being caught as pkgdb errors.)
At least separating resolution errors from implementation errors seems reasonable.

Reply:
Agreed - at this point though I don't feel like I have a good handle on how we'll want to split things up